metricbeat/module/system/users/dbus.go

```diff
@@ -0,0 +1,207 @@
+// under the License.
+
+//+build !netbsd
+
+package users
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/godbus/dbus"
+	"github.com/pkg/errors"
+)
+
+const (
+	loginObj    = "org.freedesktop.login1"
+	getAll      = "org.freedesktop.DBus.Properties.GetAll"
+	sessionList = "org.freedesktop.login1.Manager.ListSessions"
+)
+
+// sessionInfo contains useful properties for a session
+type sessionInfo struct {
+	Remote     bool
+	RemoteHost string
+	Name       string
+	Scope      string
+	Service    string
+	State      string
+	Type       string
+	Leader     uint32
+}
+
+// loginSession contains basic information on a login session
+type loginSession struct {
+	ID   uint64
+	UID  uint32
+	User string
+	Seat string
+	Path dbus.ObjectPath
+}
+
+// initDbusConnection initializes a connection to the dbus
+func initDbusConnection() (*dbus.Conn, error) {
+	conn, err := dbus.SystemBusPrivate()
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting connection to system bus")
+	}
+
+	auth := dbus.AuthExternal(strconv.Itoa(os.Getuid()))
+
+	err = conn.Auth([]dbus.Auth{auth})
+	if err != nil {
+		return nil, errors.Wrap(err, "error authenticating")
+	}
+
+	err = conn.Hello()
+	if err != nil {
+		return nil, errors.Wrap(err, "error in Hello")
+	}
+
+	return conn, nil
+}
+
+// getSessionProps returns info on a given session pointed to by path
+func getSessionProps(conn *dbus.Conn, path dbus.ObjectPath) (sessionInfo, error) {
+	busObj := conn.Object(loginObj, path)
+
+	var props map[string]dbus.Variant
+
+	err := busObj.Call(getAll, 0, "").Store(&props)
+	if err != nil {
+		return sessionInfo{}, errors.Wrap(err, "error calling DBus")
+	}
+
+	return formatSessonProps(props)
+}
+
+func formatSessonProps(props map[string]dbus.Variant) (sessionInfo, error) {
+	if len(props) < 8 {
+		return sessionInfo{}, fmt.Errorf("wrong number of fields in  info: %v", props)
+	}
+
+	remote, ok := props["Remote"].Value().(bool)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast remote to bool")
+	}
+
+	remoteHost, ok := props["RemoteHost"].Value().(string)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast remote host to string")
+	}
+
+	userName, ok := props["Name"].Value().(string)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast username to string")
+	}
+
+	scope, ok := props["Scope"].Value().(string)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast scope to string")
+	}
+
+	service, ok := props["Service"].Value().(string)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast service to string")
+	}
+
+	state, ok := props["State"].Value().(string)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast state to string")
+	}
+
+	sessionType, ok := props["Type"].Value().(string)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast type to string")
+	}
+
+	leader, ok := props["Leader"].Value().(uint32)
+	if !ok {
+		return sessionInfo{}, fmt.Errorf("failed to cast leader to uint32")
```

Comment:
Should we be more flexible when converting these properties? In case some property is not always present.

Reply:
Yah, I was thinking about this. It looks like dbus return empty objects for structures that don't have any data, so stuff like `Leader` still appears in the structure. I was thinking about doing some clever reflection, but I didn't think the readability and performance tradeoffs were worth it.